rfcs/20240722-group-query-attention/README.md

```diff
@@ -0,0 +1,215 @@
+
+  # upcast attention to fp32
+  attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(query_states.dtype)
+  attn_weights = nn.functional.dropout(attn_weights, p=self.attention_dropout, training=self.training)
+  attn_output = torch.matmul(attn_weights, value_states)
+  ```
+
+The overall workflow can be visualized as the following diagram:
+
+![GQA](GQA.png)
+
+## Proposals
+
+oneDNN's MatMul operation is used to construct the dot-product operations in
+SDPA. But the MatMul operation requires the two inputs should have the same
+batch dimension or the batch dimensions can be simply broadcasted. MQA can be
+implemented directly using typical SDPA and the broadcasting rule. But in GQA,
+the head number of Key and Value is different from the head number of Query. We
+propose two options to pre-process the Query, Key and Value tensors before
+passing them into the MatMul operations.
+
+### Option 1
+
+Add new ops and patterns. We can pre-process the Key and Value tensors following
+how they are pre-processed in the Huggingface Transformer code above. To achieve
+that, below two new operations need to be supported by oneDNN graph operation set:
+
+- Unsqueeze (see [sub-rfc document](./unsqueeze_and_expand.md))
+- Expand (see [sub-rfc document](./unsqueeze_and_expand.md))
+
+Then we would to support patterns like the above diagram.
+
+Pros:
+
+1. It makes the pattern graph to be consistent with the popular implementation
+   in the community. Frameworks can still map the framework graph directly to
+   oneDNN graph and get the fused partition.
+
+Cons:
+
+1. Need to add more operations to oneDNN graph operation set.
+2. The implementation in the community may change. Even in the Huggingface code,
+   another alternative implementation is explicitly mentioned (via
+   `torch.repeat_interleave`). Once the implementation changes, the above pros
+   will become invalid.
+
+### Option 2 (recommended)
+
+Add new patterns. We can reshape Query, Key and Values tensors from 4D to 5D and
+leverage the broadcasting semantics of MatMul operation to perform the dot-products.
+  
+1. Reshape Query from 4D shape (N, H, S, D) to 5D shape (N, G, H / G, S, D).
+2. Reshape Key from 4D shape (H, G, S, D) to 5D shape (N, G, 1, S, D).
+3. Performance 5D matmul between (N, G, H / G, S, D) and (N, G, 1, S, D). The
+   third dimension will be broadcasted from 1 to `H / G` automatically per the
+   broadcasting rule of MatMul operation.
+4. Similar reshape and broadcasting can also be applied to the dot-product of
+   Value.
+
+Here is the diagram:
+![option2](option2.png)
+
+Pros:
+
+1. No change to API.
+2. The pattern looks simpler than the pattern in option 1.
+
+Cons:
+
+1. The pattern is less intuitive from GQA definition.
+2. The pattern cannot be used to optimize a framework graph directly. Frameworks
+   will have to implement GQA fusion by themselves and leverage this option to
+   optimized the fused GQA.
```

Comment:
If this turns out to be a serious con, it would be possible to add a pass to match the Option 1 subgraph and convert it to the Option 2 subgraph, right?

Reply:
1. If it is a serious con, we need to implement option 1 adding new ops and new patterns. It is a backend implementation that matches the Option 1 subgraph and converts it to the Option 2 subgraph. We can also implement it in other ways in the backend.
2. If the pass can be done on the framework side, we only need to implement option 2.

Reply:
We will have to support and match the subgraph in Option 1 once the request pops up. With that, oneDNN will support and maintain several different patterns for the same GQA functionality. Maybe it's not an issue as even for now we choose to Option 1 as the initial step, the pattern may still change in the future as mentioned in the cons of Option 1.